src/data_structures/single_linked_list.rs

```diff
@@ -0,0 +1,248 @@
+// the public struct can hide the implementation detail
+pub struct SingleLinkedList<T> {
+    head: Link<T>,
+}
+
+type Link<T> = Option<Box<Node<T>>>;
+
+struct Node<T> {
+    elem: T,
+    next: Link<T>,
+}
+
+impl<T> SingleLinkedList<T> {
+    // Self is an alias for SingleLinkedList
+    // We implement associated function name new for single-linked-list
+    pub fn new() -> Self {
+        // for new function we need to return a new instance
+        SingleLinkedList {
+            // we refer to variants of an enum using :: the namespacing operator
+            head: None,
+        } // we need to return the variant, so there without the ;
+    }
+
+    // As we know the primary forms that self can take: self, &mut self and &self, push will change the linked list
+    // so we need &mut
+    // The push method which the signature's first parameter is self
+    pub fn push(&mut self, elem: T) {
+        let new_node = Box::new(Node {
+            elem,
+            // mem::replace function lets us steal a value out of a borrow by replacing it with another value
```

Comment:
I don't think `mem::replace` is used anywhere here (probably an old comment)

Reply:
Yes, this should be remove.